src/zos-io.cc

```diff
@@ -814,31 +897,59 @@ int __open_ascii(const char *filename, int opts, ...) {
   int fd = __open_orig(filename, opts, perms);
   int old_errno = errno;
 
+  struct file_tag *t = &sb.st_tag;
+
   if (fd >= 0) {
-    // Tag new files as ASCII (819)
+    // we need the full path
+    filename = __realpath_extended(filename, NULL);
+    // every RDONLY open we (re)determine filetags
+    if (S_ISREG(sb.st_mode) && (opts & O_RDONLY)) {
+     existing_files.erase(filename);
+     // if file ccsid doesn't match new file default or if it has the txtflag off
+     // it should be cached, except when it is FT_BINARY which always has the txtflag off
+     if (t->ft_ccsid != __ccsid_new_file() || 
+         !t->ft_ccsid == FT_BINARY && !t->ft_txtflag) {
+        if (t->ft_txtflag) {
+           existing_files.insert(std::make_pair(filename, std::make_pair(t->ft_ccsid, true)));
+        }
+        else {
+           existing_files.insert(std::make_pair(filename, std::make_pair(t->ft_ccsid, false)));
+        }            
+     }
+    }
     if (is_new_file) {
-     __tag_new_file(fd);
+     __tag_new_file_filename(fd, filename);
```

Comment:
Trying to understand this a bit more. The bool `is_new_file` is true if the file does not exist. Is there logic to handle preserving file tags for files that already exist? Let's say you open an existing file for read, process it in memory, then open it for write.

Reply:
For existing files there is only one case that needs to be handled; the heuristic choice that was made if the untagged file contains ebcdic or ascii. If it was found to contain ebcdic, then we must again enable autoconversion. For tagged (existing) files the tag is just respected asis by enhanced ascii.
If a new file matches the default for new files we don't need to do anything. In other cases we need to retrieve the cached filetags and apply them, or in case of an untagged file we need to see if it was found to contain ebcdic (txtflag in cache is 'on', or ascii (txtflag is 'off')) without actually appying a filetag or txtflag (we leave the file untagged).

Reply:
Alternative options would be:
- always assume that untagged files are ibm-1047.
- always create new files according to the chosen default and patch vim so that it respects the original filetags when it renames/recreates the file on a simple update in the editor.